Add fixed grid strategy to codebase

rkc2000 opened the pull request:
**Before submitting this PR, please make sure**:

- [ X] Your code builds clean without any errors or warnings
- [ X] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:

Add the new fixed grid strategy that was approved in the following snapshot vote:
https://snapshot.org/#/hbot-ip.eth/proposal/0xaf21d826282fcc0a4d77ad0fb775c37b7a5935af1ce106686065b4336de2f83f

Commonwealth discussion about this strategy:
https://commonwealth.im/hummingbot-foundation/discussion/3763-fixed-grid-strategy-for-hummingbot

Medium article explaining and demonstrating the strategy:
https://medium.com/@rkc2000/introducing-the-fixed-grid-strategy-for-hummingbot-d9b011c71e9b


**Tests performed by the developer**:
The new strategy has been tested out in paper trading and in live trading. Some samples from these runs are given in the medium article above. 

**Tips for QA testing**:

leastchaos reviewed (commented):
Hi, I think .cpp files should be removed as they should be compiled by the user.
You may need to add unit tests to complete the coverage of code. 
However, I think there are currently no requirements for new user-submitted strategy to have coverage. @JeremyKono @phbrgnomo can you confirm?
If so, you may add hummingbot/strategy/fixed_grid* to .coveragerc to exclude it from the check.

rkc2000 commented:
Thanks. I've removed the cpp files, and updated the .coveragerc file.

leastchaos commented on hummingbot/strategy/fixed_grid/fixed_grid.pyx:
```diff
@@ -0,0 +1,934 @@
+        return self._start_order_buy
+
+    @property
+    def base_asset(self):
```
mising type

leastchaos commented on hummingbot/strategy/fixed_grid/fixed_grid.pyx:
```diff
@@ -0,0 +1,934 @@
+        return self._market_info.base_asset
+
+    @property
+    def quote_asset(self):
```
missing type

leastchaos commented on hummingbot/strategy/fixed_grid/fixed_grid.pyx:
```diff
@@ -0,0 +1,934 @@
+        return self._market_info.quote_asset
+
+    @property
+    def trading_pair(self):
```
missing type

leastchaos commented on hummingbot/strategy/fixed_grid/fixed_grid.pyx:
```diff
@@ -0,0 +1,934 @@
+
+        return pd.DataFrame(data=data, columns=columns)
+
+    def market_status_data_frame(self, market_trading_pair_tuples: List[MarketTradingPairTuple]) -> pd.DataFrame:
```
Is there any difference between the strategy_base market_status_dataframe and this market_status_data_frame?

leastchaos commented on hummingbot/strategy/fixed_grid/fixed_grid_order_tracker.pyx:
```diff
@@ -0,0 +1,49 @@
+NaN = float("nan")
+
+
+cdef class FixedGridOrderTracker(OrderTracker):
```
the method added seems to already be available in the OrderTracker class. I think it may be better to refer to OrderTracker if there is no difference.

rkc2000 replied:
Thanks. I've now added the missing types, and used the existing market_status_data_frame and OrderTracker. I recompiled and tested the code to make sure everything works as before.

leastchaos replied:
Thanks. Nothing else from me. although a bit weird to see the fix to bitfinex connector and avell strategy in this pr as it is not part of the scope of this pr.

rkc2000 replied:
I had to make those fixes to get the code to compile, so I could test it. I can remove those changes for the sake of this merge if necessary.

leastchaos replied:
Hi, I received information that any new functionality submitted will need to include the necessary unit tests to provide the code coverage. So you have to remove the line added in .coveragerc and add the unit test required in order for this strategy to be merged to the codebase. You may refer to the tests folder on how to add the unit tests.
Also, please resolve the merge conflict in the pr. Thank you.

- Tech Review DAO

leastchaos reviewed (commented):
minor comments.

rapcmia commented:
PR testing in progress, Just need to add some adjustments on UI since its missing `>>>` character when we are updating the `grid_price_ceiling` and `grid_price_floor`

![image](https://user-images.githubusercontent.com/73840223/166200965-ac01fd3d-88f4-4ae3-898e-4f0e0e193768.png)

Also I would like to some help for the equation of `price at i-th level..` where can we find the i value? Thank you!
![image](https://user-images.githubusercontent.com/73840223/166201558-341df601-ed0a-44b2-82cd-415c846b5c37.png)

rkc2000 commented:
Thanks. I have added the missing >>> characters. The i-th level during runtime is given by the "Current grid level" that is displayed in the status frame. I'm not sure if this answers the question, please let me know if I can clarify. 

P.S. I am working on adding unit tests for code coverage, I will have that done soon. 

<img width="668" alt="Screenshot 2022-05-02 at 17 34 51" src="https://user-images.githubusercontent.com/6465285/166281527-8f0b8589-cbcb-49fa-a593-26629f9cde5c.png">

rapcmia commented:
Hi @rkc2000 the latest commit added the missing `>>>` thanks!

> PR testing in progress, Just need to add some adjustments on UI since its missing `>>>` character when we are updating the `grid_price_ceiling` and `grid_price_floor`

Ahh yes we are trying to create a calc to monitor how the strategy creates order when the grid operation starts. Basically, we want to find out how the bot creates the order levels for buy and sell orders on the screenshot below:
![image](https://user-images.githubusercontent.com/73840223/166430564-21d41e96-5d9f-42c6-898d-0d3b07bca331.png)

> The i-th level during runtime is given by the "Current grid level" that is displayed in the status frame. I'm not sure if this answers the question, please let me know if I can clarify.

JeremyKono commented:
Hi @rkc2000, please resolve conflicts in this PR. Thanks!

rkc2000 commented:
I've resolved the conflict, thanks!

aarmoa commented:
> Hi, I received information that any new functionality submitted will need to include the necessary unit tests to provide the code coverage. So you have to remove the line added in .coveragerc and add the unit test required in order for this strategy to be merged to the codebase. You may refer to the tests folder on how to add the unit tests. Thank you.
> 
> * Tech Review DAO

Hello @rkc2000. As mentioned by @leastchaos please remove the changes in the `.coveragerc` file. You should not add an exception for the new code and instead you have to add the required unit tests for the new functionality.

rkc2000 commented:
@aarmoa @leastchaos I've now added unit tests for this new strategy (and removed the changes in .coveragerc). Sorry about the delay, it took me a while to figure out the logic. Do let me know if anything further is required.

alpha5005 commented:
Is this strategy testes and are we waiting for maintainer to approve? Very excited to test this out.

leastchaos commented:
@JeremyKono Hi, please approve the test workflow

leastchaos commented:
![image](https://user-images.githubusercontent.com/64965782/168923523-a733598e-4b03-462e-8ae0-3af255aec77a.png)
![image](https://user-images.githubusercontent.com/64965782/168923580-6ab83b58-d268-4c86-9d3e-b0fa0b9de069.png)

Hi you may fix according to the details above. you can view the above by clicking details of the ci.

rkc2000 commented:
I've made the fixes. Could you please try the checks again? Thanks.

abduramann commented:
When will this be added to the new release? Thanks.

alpha5005 commented:
@rkc2000 I am testing your fixed grid strategy.
I am noticing following behavior:
I am observing that every time my limit maker order is filled all other orders are cancelled.  Then it automatically put all orders again but when one of those orders get filled it again cancels all other orders and this keeps on repeating in indefinitely upon every limit maker order is filled. I am not sure it is bug in code but it is definitely improvement you may like to consider as for grid strategies we do not want constant opening and cancelling of 100s levels on exchange if any order is being filled. This also creates too many pings to exchange, imposing API rate limits. 

I may be wrong but could you please help me to understand why it might be cancelling my all other order once a single limit marker order gets filled and then it puts new orders again?

rkc2000 commented:
@alpha5005 are you using the latest version of the code? I have already made this change. When an order is filled, only the level above/below the fill gets updated. The other orders remain untouched. I specifically did this since I was hitting rate limits when orders are cancelled and reposted.

rkc2000 commented:
@leastchaos looks like the CI is still failing, but I can’t see any errors in the log now. I had already fixed all the errors/warnings it was showing last time. Could you help with finding what the issue is?

rkc2000 commented:
@alpha5005 see this commit https://github.com/hummingbot/hummingbot/pull/5249/commits/ee68d7f729e109c72df3f63512a5440b90a7b27e

leastchaos commented:
you can try run the following code and then commit the changes it performed.
'conda activate hummingbot && pre-commit run --files $(git diff --name-only origin/development)'
make sure your development branch is also updated with the latest version of development of the official repo.

alpha5005 commented:
> @alpha5005 are you using the latest version of the code? I have already made this change. When an order is filled, only the level above/below the fill gets updated. The other orders remain untouched. I specifically did this since I was hitting rate limits when orders are cancelled and reposted.

I think that could be the problem. Let me update my local branch and see what happens

alpha5005 commented:
> > @alpha5005 are you using the latest version of the code? I have already made this change. When an order is filled, only the level above/below the fill gets updated. The other orders remain untouched. I specifically did this since I was hitting rate limits when orders are cancelled and reposted.
> 
> I think that could be the problem. Let me update my local branch and see what happens.

I think I am getting confuse how to get update code having this - "The full grid is not refreshed anymore when a new order is filled" committed changes. May be I am doing wrong repository and branch clone or pulling stale PR.

Could you please suggest a sequence of git commands from beginning so that I can pull this changes?

alpha5005 commented:
@rkc2000 which branch should I clone ? Here is what I am seeing on various branches:
rkc2000/hummingbot/Master branch has c_did_complete_buy_order with code self._create_timestamp = self._current_timestamp + self._order_refresh_time
rkc2000/hummingbot/Master branch has c_did_complete_sell_order with code         self._create_timestamp = self._current_timestamp + self._filled_order_delay

rkc2000/hummingbot/feat/fixed_grid_strategy has c_did_complete_buy_order with code self._create_timestamp = self._current_timestamp + self._filled_order_delay
rkc2000/hummingbot/feat/fixed_grid_strategy has c_did_complete_sell_order with code self._create_timestamp = self._current_timestamp + self._filled_order_delay

JeremyKono commented:
@rkc2000 you have until end of May to do the fixes given by leastchaos or it will not be added to the 1.5 release. Thanks!

leastchaos commented:
Hi, I add some step here

1. Please update/rebase your local branch to the latest development version. 
2. run the following to update the code to the proper style
`conda activate hummingbot && pre-commit run --files $(git diff --name-only origin/development)'`
3. run the following to check your code coverage and test is ok
```
make run_coverage
make development-diff-cover
```
if you cannot run the above the following commands should also work fine
```
coverage run -m nose --exclude-dir="test/connector" --exclude-dir="test/debug" --exclude-dir="test/mock"
coverage xml
diff-cover --compare-branch=origin/development coverage.xml
```
I have received this error in your test. on checking, the module have been moved to hummingbot.connector.mock.mock_paper_exchange
`  File "hummingbot/strategy/fixed_grid/test_fixed_grid.py", line 17, in <module>
    from test.mock.mock_paper_exchange import MockPaperExchange`
![image](https://user-images.githubusercontent.com/64965782/170615269-839ea1a9-ea14-4f3e-87a2-2c99296a503e.png)

4. once all tests and the coverage in point 3 are passed, you should be able to merge this pr without further issue.

rkc2000 commented:
Hi @leastchaos I've been trying to do as you suggested but I'm running into some issues. The command 'pre-commit run --files $(git diff --name-only origin/development)' is throwing a lot of errors. Any ideas on how to fix these? 
<img width="1090" alt="Screenshot 2022-05-27 at 23 19 07" src="https://user-images.githubusercontent.com/6465285/170764223-1fdbbbe2-91ab-4d99-a1b0-519e8eed4725.png">
<img width="765" alt="Screenshot 2022-05-27 at 23 19 21" src="https://user-images.githubusercontent.com/6465285/170764235-85b3f474-f4cc-4659-a5c6-52b80e9e3ac9.png">

leastchaos commented:
I am not sure why but your latest git merge is adding a lot of unrelated changes into this branch. that's why you are seeing a lot of errors.
I am not that familiar with using git for merge conflicts so if it was me, usually if the git pull or git pull --rebase result in a lot of merge conflict or unrelated changes appearing like your case and they are well separated since you are only adding new files, I would 
1. copy my changes out to a temp folder
2. hard reset to latest development branch
`git reset --hard upstream development`
3. copy the changes back in then commit
4. force push the latest commit to the branch
`git push -f`
this would result only in 1 commit its the easiest way i can think of but probably not the best way.

@cardosofede  can you help?

cardosofede commented:
Hi @leastchaos, ping me on discord and I will help you with the GitHub stuff.
My handle is dardonacci#4056

leastchaos commented:
@rkc2000 please contact @cardosofede to fix the issue. Thanks.

leastchaos commented:
@rkc2000 i suspect your local "development" branch is not updated
You can probably do the following.

update your local development branch
```
git checkout development
git fetch upstream development
git pull upstream development
```
reset your branch to before pre-commit run
```
git checkout <your branch>
git reset --hard HEAD~1
```
run pre-commit which runs on files that were changed as compared to your development branch
```
conda activate hummingbot && pre-commit run --files $(git diff --name-only origin/development)
git push -f
```
run local tests, the following code runs the same code as the ci checks so if this pass, you should likely pass here
```
coverage run -m nose --exclude-dir="test/connector" --exclude-dir="test/debug" --exclude-dir="test/mock"
coverage xml
diff-cover --compare-branch=origin/development coverage.xml
```
clean up test and coverage (if any) and commit if required

abduramann commented:
Hey @rkc2000 ,

When I clone your repo via `git clone --branch feat/fixed_grid_strategy https://github.com/rkc2000/hummingbot.git` and try to run it, I'm getting this error:

Traceback (most recent call last):
  File "bin/hummingbot.py", line 11, in <module>
    from hummingbot.client.config.config_helpers import create_yml_files, read_system_configs_from_yml, write_config_to_yml
  File "/home/abduraman/crypto/humsrc/hummingbot/hummingbot/client/config/config_helpers.py", line 15, in <module>
    from hummingbot.client.config.global_config_map import global_config_map
  File "/home/abduraman/crypto/humsrc/hummingbot/hummingbot/client/config/global_config_map.py", line 13, in <module>
    from hummingbot.core.rate_oracle.rate_oracle import RateOracle, RateOracleSource
  File "/home/abduraman/crypto/humsrc/hummingbot/hummingbot/core/rate_oracle/rate_oracle.py", line 10, in <module>
    from hummingbot.connector.exchange.ascend_ex.ascend_ex_api_order_book_data_source import AscendExAPIOrderBookDataSource
  File "/home/abduraman/crypto/humsrc/hummingbot/hummingbot/connector/exchange/ascend_ex/ascend_ex_api_order_book_data_source.py", line 12, in <module>
    from hummingbot.connector.exchange.ascend_ex import ascend_ex_constants as CONSTANTS
  File "/home/abduraman/crypto/humsrc/hummingbot/hummingbot/connector/exchange/ascend_ex/ascend_ex_constants.py", line 3, in <module>
    from hummingbot.core.data_type.in_flight_order import OrderState
  File "/home/abduraman/crypto/humsrc/hummingbot/hummingbot/core/data_type/in_flight_order.py", line 12, in <module>
    from hummingbot.core.data_type.limit_order import LimitOrder
ModuleNotFoundError: No module named 'hummingbot.core.data_type.limit_order'

abduramann commented:
Never mind @rkc2000 . I solved the problem doing below. However, note that `conf_fixed_grid_strategy_TEMPLATE.yml ` file does NOT exist in feat/fixed_grid_strategy branch. I had to retrieve it from master branch.

```
git init hummingbot
cd hummingbot
git remote add origin https://github.com/hummingbot/hummingbot
git fetch --prune --progress --no-recurse-submodules origin +refs/heads/*:refs/remotes/origin/* +refs/tags/*:refs/tags/* +504f669ab95603ae5e6c7ad11bdfa5cac77cca3e:refs/remotes/pull/5249/merge
git checkout --progress --force refs/remotes/pull/5249/merge
```

rkc2000 commented:
Thanks @leastchaos for all the help. I’ll try to fix it later today, also with the help of @cardosofede

rkc2000 commented:
@abduramann thanks for your pointers as well

JeremyKono commented:
Hi @rkc2000 please fix the failed tests on this PR.

rkc2000 commented:
Hi @JeremyKono im trying to and am in the final stages. Will complete within a day or 2 max

rkc2000 commented:
Hi, I think I've fixed all the issues now. I wrote the code back in Feb and there have been some subtle changes to the codebase since then, so it took a while to identify and sort out some minor snags! I had messed up this branch by desperately trying out some stuff, so I created a new clean branch to do the merging:
https://github.com/hummingbot/hummingbot/pull/5371

I hope this works. Thanks @leastchaos and @cardosofede for all your help!

aarmoa commented:
@cardosofede @JeremyKono this PR has been replaced by a new one and should be closed.

JeremyKono commented:
Will be closing this PR since a new PR has been opened to address the issues. https://github.com/hummingbot/hummingbot/pull/5371
